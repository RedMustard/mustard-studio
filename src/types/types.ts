export type KeyboardKeyCount = 88 | 61 | 49 | 37 | 25;
export type KeyboardKeyColor = 'white' | 'black';
export type Octave = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;
export type OscillatorId = 'osc1' | 'osc2' | 'oscSub';
export type ValueUnit = 'percent' | 'cent' | 'pan';


export type StudioService = {
    settings: {
        master: {
            volume: number,
            pan: number,
        },
        osc1: OscillatorSettings & OscillatorDetune,
        osc2: OscillatorSettings & OscillatorDetune,
        oscSub: OscillatorSettings,
    },
    gainNodes: {
        master: GainNode,
        osc1: GainNode,
        osc2: GainNode,
        oscSub: GainNode,
    },
    panNodes: {
        master: StereoPannerNode,
        osc1: StereoPannerNode,
        osc2: StereoPannerNode,
        oscSub: StereoPannerNode,
    },
};

export type OscillatorSettings = {
    enabled: boolean,
    octave: number,
    pan: number,
    type: OscillatorType,
    volume: number,
};

export type OscillatorDetune = {
    detune: number,
};

export type StudioServiceActionType =
    'SET_MASTER_VOLUME' |
    'SET_MASTER_GAIN' |
    'SET_MASTER_PAN_NODE' |
    'SET_MASTER_PAN_POSITION' |
    'SET_OSC_1_VOLUME' |
    'SET_OSC_2_VOLUME' |
    'SET_OSC_SUB_VOLUME' |
    'SET_OSC_1_GAIN' |
    'SET_OSC_2_GAIN' |
    'SET_OSC_SUB_GAIN' |
    'SET_OSC_1_ENABLED' |
    'SET_OSC_2_ENABLED' |
    'SET_OSC_SUB_ENABLED' |
    'SET_OSC_1_TYPE' |
    'SET_OSC_2_TYPE' |
    'SET_OSC_SUB_TYPE' |
    'SET_OSC_1_DETUNE' |
    'SET_OSC_2_DETUNE' |
    'SET_OSC_1_PAN_NODE' |
    'SET_OSC_2_PAN_NODE' |
    'SET_OSC_SUB_PAN_NODE' |
    'SET_OSC_1_PAN_POSITION' |
    'SET_OSC_2_PAN_POSITION' |
    'SET_OSC_SUB_PAN_POSITION' |
    'SET_OSC_1_OCTAVE' |
    'SET_OSC_2_OCTAVE' |
    'SET_OSC_SUB_OCTAVE';

export type StudioServiceAction = {
    type: StudioServiceActionType;
    payload: any;
};

export type DispatchFunction = (args: StudioServiceAction) => void;

