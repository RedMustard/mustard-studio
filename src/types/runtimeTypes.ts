export enum OscillatorTypeEnum {
    SINE = 'sine',
    SAWTOOTH = 'sawtooth',
    TRIANGLE = 'triangle',
    SQUARE = 'square',
    CUSTOM = 'custom',
}
export const OscillatorTypes = Object.values(OscillatorTypeEnum) as OscillatorType[];

export enum FilterTypeEnum {
    LOWPASS = 'lowpass',
    HIGHPASS = 'highpass',
    BANDPASS = 'bandpass',
    LOWSHELF = 'lowshelf',
    HIGHSHELF = 'highshelf',
    PEAKING = 'peaking',
    NOTCH = 'notch',
    ALLPASS = 'allpass',
}
export const FilterTypes = Object.values(FilterTypeEnum) as BiquadFilterType[];


export enum ValueUnitEnum {
    PERCENT = 'percent',
    CENT = 'cent',
    PAN = 'pan',
    SECONDS = 'seconds',
}

export enum OscillatorIdEnum {
    OSC_1 = 'osc1',
    OSC_2 = 'osc2',
    OSC_SUB = 'oscSub',
}
