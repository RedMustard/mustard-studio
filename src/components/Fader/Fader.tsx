import { h } from 'preact';
import { useState } from 'preact/hooks';
import { ValueUnitEnum } from '../../types/runtimeTypes';
import { ValueUnit } from '../../types/types';


interface FaderProps {
    classSuffix: string;
    maxValue: number;
    minValue: number;
    onInput: (value: number) => void;
    stepResolution: number;
    value: number;
    valueUnit: ValueUnit;
}

export const Fader = ({
    classSuffix,
    maxValue,
    minValue,
    onInput,
    stepResolution,
    value,
    valueUnit,
}: FaderProps) => {
    const [showValue, setShowValue] = useState(false);

    const getPrettifiedValue = () => {
        switch (valueUnit) {
            case ValueUnitEnum.PERCENT:
                return `${Math.floor(value * 100)}%`;
            case ValueUnitEnum.CENT:
            default:
                return value;
        }
    };

    return (
        <div className={`fader  ${classSuffix}__fader`}>
            {
                showValue
                    ? (<span className="fader__value">{getPrettifiedValue()}</span>)
                    : null
            }
            <input
                title={getPrettifiedValue().toString()}
                onInput={(event) => onInput(parseFloat(event.currentTarget.value))}
                onMouseDown={() => setShowValue(true)}
                onMouseUp={() => setShowValue(false)}
                list={`${classSuffix}-fader`}
                max={maxValue}
                min={minValue}
                name={`${classSuffix}-fader`}
                step={stepResolution}
                type="range"
                value={value}
            />
            <datalist id={`${classSuffix}-fader`}>
                <option value={maxValue} label="Mute" />
                <option value={minValue} label="100%" />
            </datalist>
        </div>
    );
};
