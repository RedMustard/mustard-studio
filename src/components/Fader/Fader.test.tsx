import { h } from 'preact';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Fader } from './Fader';
import { ValueUnitEnum } from '../../types/runtimeTypes';


const baseProps = {
    classSuffix: 'foo',
    maxValue: 1,
    minValue: 0,
    onInput: jest.fn(),
    onMouseDown: jest.fn(),
    stepResolution: 0.1,
    value: 0.5,
    valueUnit: ValueUnitEnum.CENT,
};

describe('<FaderKey />', () => {
    it('renders with basic props', () => {
        const wrapper = shallow(
            <Fader
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with mousedown', () => {
        const wrapper = shallow(
            <Fader
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown');
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with mousedown and mouseup', () => {
        const wrapper = shallow(
            <Fader
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown');
        wrapper.find('input').simulate('mouseup');
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with PERCENT valueUnit', () => {
        const wrapper = shallow(
            <Fader
                {...baseProps}
                valueUnit={ValueUnitEnum.PERCENT}
            />,
        );
        wrapper.find('input').simulate('mousedown');
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with PAN valueUnit onMouseDown', () => {
        const wrapper = shallow(
            <Fader
                {...baseProps}
                valueUnit={ValueUnitEnum.PAN}
            />,
        );
        wrapper.find('input').simulate('mousedown');
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('fires onInput', () => {
        const wrapper = mount(
            <Fader
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('input');
        expect(baseProps.onInput).toBeCalled();
    });

    it('fires onMouseDown', () => {
        const wrapper = mount(
            <Fader
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown');
        expect(baseProps.onMouseDown).toBeCalled();
    });
});
