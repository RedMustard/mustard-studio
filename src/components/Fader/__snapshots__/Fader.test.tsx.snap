// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FaderKey /> renders with PAN valueUnit onMouseDown 1`] = `
<div
  className="fader  foo__fader"
>
  <span
    className="fader__value"
  >
    5
  </span>
  <input
    list="foo-fader"
    max={1}
    min={0}
    name="foo-fader"
    onInput={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    step={0.1}
    title="5"
    type="range"
    value={0.5}
  />
  <datalist
    id="foo-fader"
  >
    <option
      label="Mute"
      value={1}
    />
    <option
      label="100%"
      value={0}
    />
  </datalist>
</div>
`;

exports[`<FaderKey /> renders with PERCENT valueUnit 1`] = `
<div
  className="fader  foo__fader"
>
  <span
    className="fader__value"
  >
    50%
  </span>
  <input
    list="foo-fader"
    max={1}
    min={0}
    name="foo-fader"
    onInput={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    step={0.1}
    title="50%"
    type="range"
    value={0.5}
  />
  <datalist
    id="foo-fader"
  >
    <option
      label="Mute"
      value={1}
    />
    <option
      label="100%"
      value={0}
    />
  </datalist>
</div>
`;

exports[`<FaderKey /> renders with SECONDS valueUnit onMouseDown 1`] = `
<div
  className="fader  foo__fader"
>
  <span
    className="fader__value"
  >
    0.5s
  </span>
  <input
    list="foo-fader"
    max={1}
    min={0}
    name="foo-fader"
    onInput={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    step={0.1}
    title="0.5s"
    type="range"
    value={0.5}
  />
  <datalist
    id="foo-fader"
  >
    <option
      label="Mute"
      value={1}
    />
    <option
      label="100%"
      value={0}
    />
  </datalist>
</div>
`;

exports[`<FaderKey /> renders with basic props 1`] = `
<div
  className="fader  foo__fader"
>
  <input
    list="foo-fader"
    max={1}
    min={0}
    name="foo-fader"
    onInput={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    step={0.1}
    title="0.5"
    type="range"
    value={0.5}
  />
  <datalist
    id="foo-fader"
  >
    <option
      label="Mute"
      value={1}
    />
    <option
      label="100%"
      value={0}
    />
  </datalist>
</div>
`;

exports[`<FaderKey /> renders with mousedown 1`] = `
<div
  className="fader  foo__fader"
>
  <span
    className="fader__value"
  >
    0.5
  </span>
  <input
    list="foo-fader"
    max={1}
    min={0}
    name="foo-fader"
    onInput={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    step={0.1}
    title="0.5"
    type="range"
    value={0.5}
  />
  <datalist
    id="foo-fader"
  >
    <option
      label="Mute"
      value={1}
    />
    <option
      label="100%"
      value={0}
    />
  </datalist>
</div>
`;

exports[`<FaderKey /> renders with mousedown and mouseup 1`] = `
<div
  className="fader  foo__fader"
>
  <input
    list="foo-fader"
    max={1}
    min={0}
    name="foo-fader"
    onInput={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    step={0.1}
    title="0.5"
    type="range"
    value={0.5}
  />
  <datalist
    id="foo-fader"
  >
    <option
      label="Mute"
      value={1}
    />
    <option
      label="100%"
      value={0}
    />
  </datalist>
</div>
`;
